### PR TITLE
AP_Frsky_Telem: added passthrough frame 0x500B with height above terrain

### PR DESCRIPTION
**Pull request by yaapu**
This adds a new frame 0x500B with height above terrain, it's sent only when terrain is enabled.

### CODE REVIEW

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -4635,6 +4637,89 @@ void GCS_MAVLINK::send_generator_status() const
...
+
+    if (home_is_set) {
+        altitude_relative = global_position_int_relative_alt() * 0.001; //meters above home
+    } else {
+        // fall back to local position
+        altitude_relative = altitude_local;
+    }
+
+#if AP_TERRAIN_AVAILABLE
+    AP_Terrain &terrain = AP::terrain();
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp:4694]: `const` (if we can)

**Review by peterbarker — CHANGES_REQUESTED** [review]
Looking really good!

File: ArduCopter/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -483,6 +483,7 @@ static const ap_message STREAM_EXTRA3_msgs[] = {
     MSG_ESC_TELEMETRY,
     MSG_GENERATOR_STATUS,
     MSG_WINCH_STATUS,
+    MSG_ALTITUDE,
```
[the remark this answers is not in the document]
Comment (yaapu) [reply ArduCopter/GCS_Mavlink.cpp:486]: ok, makes sense, fixed

File: Tools/autotest/common.py [hunk 3]
```diff
@@ -9209,6 +9210,22 @@ def tfp_validate_rpm(self, value):
             return False
         return True
 
+    def tfp_validate_terrain(self, value):
+        self.progress("validating terrain(0x%02x)" % value)
+        alt_above_terrain_dm = self.bit_extract(value, 2, 10) * (10 ^ self.bit_extract(value, 0, 2)) * 0.1 * (self.bit_extract(value, 12, 1) == 1 and -1 or 1)  # noqa
```
[the remark this answers is not in the document]
Comment (yaapu) [reply Tools/autotest/common.py:9215]: you mean a structure or class to incapsulate passthrough "wild" encoding?

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 4]
```diff
@@ -852,6 +853,7 @@ ap_message GCS_MAVLINK::mavlink_id_to_ap_message_id(const uint32_t mavlink_id) c
         { MAVLINK_MSG_ID_EFI_STATUS,            MSG_EFI_STATUS},
         { MAVLINK_MSG_ID_GENERATOR_STATUS,      MSG_GENERATOR_STATUS},
         { MAVLINK_MSG_ID_WINCH_STATUS,          MSG_WINCH_STATUS},
+        { MAVLINK_MSG_ID_ALTITUDE,          MSG_ALTITUDE},
```
[the remark this answers is not in the document]
Comment (yaapu) [reply libraries/GCS_MAVLink/GCS_Common.cpp:856]: ok fixed

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 5]
```diff
@@ -4635,6 +4637,89 @@ void GCS_MAVLINK::send_generator_status() const
...
+    bool got_position = false;
+    bool home_is_set = false;
+    bool got_relative_position_ned = false;
+
+    float altitude_monotonic;
+    float altitude_amsl;
+    float altitude_local;
+    float altitude_relative;
+    float altitude_terrain;
+    float bottom_clearance;
```
[the remark this answers is not in the document]
Comment (yaapu) [reply libraries/GCS_MAVLink/GCS_Common.cpp:4654]: will do

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 6]
```diff
@@ -4635,6 +4637,89 @@ void GCS_MAVLINK::send_generator_status() const
...
+
+    {
+        AP_Baro &_baro = AP::baro();
+        WITH_SEMAPHORE(_baro.get_semaphore());
```
[the remark this answers is not in the document]
Comment (yaapu) [reply libraries/GCS_MAVLink/GCS_Common.cpp:4658]: yep, thanks

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 7]
```diff
@@ -4635,6 +4637,89 @@ void GCS_MAVLINK::send_generator_status() const
...
+
+        got_position = _ahrs.get_position(loc);
+        home_is_set = _ahrs.home_is_set();
+        got_relative_position_ned = _ahrs.get_relative_position_NED_origin(local_position);
+    }
+
+    if (got_position) {
+        altitude_amsl = global_position_int_alt() * 0.001; //meters above ground/sea level
+    } else {
+        // fall back to baro altitude
```
[the remark this answers is not in the document]
Comment (yaapu) [reply libraries/GCS_MAVLink/GCS_Common.cpp:4675]: ok, so my solution is to wait for ahrs to have a position, no fall back

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 8]
```diff
@@ -4635,6 +4637,89 @@ void GCS_MAVLINK::send_generator_status() const
...
+    AP_Terrain &terrain = AP::terrain();
+    if (terrain.enabled()) {
+        if (!terrain.height_above_terrain(altitude_terrain, true)) {
+            altitude_terrain = -1000;   //mavlink specs value for unknown
+        }
+    }
+#endif
+
+    bottom_clearance = -1;  //mavlink specs value for unavailable
+    RangeFinder *rangefinder = RangeFinder::get_singleton();
```
[the remark this answers is not in the document]
Comment (yaapu) [reply libraries/GCS_MAVLink/GCS_Common.cpp:4703]: fixed

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 1 above)
Comment (yaapu) [reply libraries/GCS_MAVLink/GCS_Common.cpp:4694]: fixed

File: libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp [hunk 9]
```diff
@@ -200,6 +202,15 @@ bool AP_Frsky_SPort_Passthrough::is_packet_ready(uint8_t idx, bool queue_empty)
             packet_ready = rpm->num_sensors() > 0;
         }
         break;
+    case TERRAIN:
+        {
+            packet_ready = false;
+#if AP_TERRAIN_AVAILABLE
```
[the remark this answers is not in the document]
Comment (yaapu) [reply libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:208]: ...now I did :-)

[the remark this answers is not in the document]
Comment (yaapu) [reply libraries/AP_Frsky_Telem/AP_Frsky_SPort_Passthrough.cpp:208]: but on real hardware for undefining it breaks SITL

```
[632/833] Compiling libraries/SITL/SIM_Airspeed_DLVR.cpp
../../libraries/SITL/SIM_Aircraft.cpp: In constructor ‘SITL::Aircraft::Aircraft(const char*)’:
../../libraries/SITL/SIM_Aircraft.cpp:66:20: error: ‘terrain’ is not a member of ‘AP’
     terrain = &AP::terrain();

```

**Comment by tridge** [comment]
we can use TERRAIN_STATUS message to test terrain data

**Comment by yaapu** [comment]
> we can use TERRAIN_REPORT message to test terrain data

OK, updated autotest to use TERRAIN_REPORT.

**Comment by peterbarker** [comment]
Merged, thanks!

Apologies for the whole `ALTITUDE` thing; I'd forgotten we reported terrain-alt this way.

**Comment by yaapu** [comment]
> Apologies for the whole `ALTITUDE` thing; I'd forgotten we reported terrain-alt this way.

no worries, thanks for merging

**Pull request merged**